spec/models/user_statistics_spec.rb

```diff
@@ -0,0 +1,96 @@
+#  bucket                        :tsrange          not null, primary key
+#  user_id                       :bigint           primary key
+#
+# Indexes
+#
+#  constraint_baw_user_statistics_non_overlapping  (user_id,bucket) USING gist
+#  constraint_baw_user_statistics_unique           (user_id,bucket) UNIQUE
+#  index_user_statistics_on_user_id                (user_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (user_id => users.id)
+#
+
+RSpec.describe UserStatistics, type: :model do
+  subject { FactoryBot.build(:user_statistics) }
+
+  it 'has a valid factory' do
+    expect(FactoryBot.create(:user_statistics)).to be_valid
+  end
+
+  it { is_expected.to belong_to(:user).optional(true) }
+
+  its('class.primary_key') { is_expected.to eq(['user_id', 'bucket']) }
+
+  it {
+    is_expected.to validate_numericality_of(:audio_original_download_count)
+      .is_greater_than_or_equal_to(0)
+      .only_integer
+  }
+
+  it {
+    is_expected.to validate_numericality_of(:audio_segment_download_count).is_greater_than_or_equal_to(0).only_integer
+  }
+
+  it { is_expected.to validate_numericality_of(:audio_download_duration).is_greater_than_or_equal_to(0) }
+
+  context 'with data' do
+    create_audio_recordings_hierarchy
+
+    it_behaves_like 'a model with a temporal stats bucket', {
+      model: UserStatistics,
+      parent_factory: :user,
+      parent: :reader_user,
+      other_key: :user_id
+    }
+
+    context 'when inserting stats' do
+      it 'can increment original download count' do
+        UserStatistics.increment_original(reader_user, audio_recording)
+
+        stats = UserStatistics.first
+        expect(stats.audio_original_download_count).to eq 1
+        expect(stats.audio_download_duration).to eq audio_recording.duration_seconds
+
+        UserStatistics.increment_original(reader_user, audio_recording)
+        stats.reload
+        expect(stats.audio_original_download_count).to eq 2
+        expect(stats.audio_download_duration).to eq(audio_recording.duration_seconds * 2)
+
+        reader_user.reload_statistics
+        actual = reader_user.statistics
+        expect(actual.audio_original_download_count).to eq 2
+        expect(actual.audio_download_duration).to eq(audio_recording.duration_seconds * 2)
+      end
+
+      it 'can increment segment download count' do
+        UserStatistics.increment_segment(reader_user, duration: 30)
+
+        stats = UserStatistics.first
+        expect(stats.audio_segment_download_count).to eq 1
+        expect(stats.audio_download_duration).to eq 30.0
+
+        UserStatistics.increment_segment(reader_user, duration: 12.5)
+
+        stats.reload
+        expect(stats.audio_segment_download_count).to eq 2
+        expect(stats.audio_download_duration).to eq 42.5
+
+        reader_user.reload_statistics
+        actual = reader_user.statistics
+        expect(actual.audio_segment_download_count).to eq 2
+        expect(actual.audio_download_duration).to eq 42.5
+      end
+    end
+  end
```

Comment:
> A lot of this is the same as the audio recordings spec file. Potentially worth refactoring in the inevitable even of

Valid point. I did do this with the shared examples for things that should behave the same (see ` it_behaves_like 'a model with a temporal stats bucket'`).

For the ones not in the shared examples: each field was a different name, and there are some other subtle differences in the summations, and abilities in the users (anon can get segments but not original files). Net result is not worth refactoring because the amount of noise.

> It may also be worth stress testing how accurate these numbers are (depending on how important we think that is). Such as a test with large numbers (ie. testing that bigint is functional), or testing that irrational binary numbers are handled properly by ruby (ie. 0.1 incremented 100 times).

I looked into this. Because we're using postgres we have transactional guarantees on races. Because we're only incrementing (not setting to an absolute value) it also does not matter which order the operations occur in. We're effectively only limited by database connection pool and shared memory buffers. It is theoretically possible to deadlock, but the `UNLOGGED` statement makes it far less likely. The insert/updates occurs in <1ms in testing so I'm happy it's good enough.